### DISCUSSION TITLE
Save audio and video tracks from rtsp to .ts files

### DISCUSSION THREAD

**Discussion by jackovka** [bluenviron/gortsplib#230] [Questions] [answered]
Hello!

I catch an RTSP stream with audio and video tracks from the server and save it to .ts files. There were no problems with the video with the H264 codec, following the “client-read-format-h264-save-to-disk” example, I form packets and write them to a file. But I can't figure out how to do this with an AAC codec audio track (defined as MPEG4Audio) and how to sync it with the video.

Similarly, I tried to create a stream using AddElementaryStream, ElementaryPID=255, StreamType=astits.StreamTypeAACAudio, and there is a track in the file, but its codec in VLC (ADTS) is different from the original one (MPEG AAC Audio).

Tell me, please, how can I save both tracks to files?

**Comment by aler9** [comment, marked as answer]
Hello, a full MPEG-TS muxer is available in mediacommon:
https://github.com/bluenviron/mediacommon/blob/main/pkg/formats/mpegts/writer.go

In particular:
* MPEG4-audio units must be wrapped inside ADTS packets
* you have to subtract initial audio DTS from audio DTS such that the first audio packet has DTS = 0
* you have to subtract initial video DTS from video DTS and video PTS such that the first video packet has PTS = DTS = 0

**Comment by jackesdavid** [comment]
is there any example on how to save audio and video in the same .ts file ?
I really appreciate any help

**Reply by aler9** [reply]
i've added the client-play-format-h264-mpeg4audio-save-to-disk that shows how to do that:
https://github.com/bluenviron/gortsplib/blob/main/examples/client-play-format-h264-mpeg4audio-save-to-disk/main.go
